### ISSUE TITLE
param.DataFrame not accepting input

### ISSUE THREAD

**Issue by kcpevey** [holoviz/param#292]
Maybe I'm not calling this correctly? 

`foo = param.DataFrame(columns=3)`
results in:  
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-526280bfed91> in <module>
      5 # class test(param.Parameterized):
      6 #     df = param.DataFrame(rows=3, columns=4)
----> 7 foo = param.DataFrame(columns=3)

~/ers/gitHub/param/master/param/param/__init__.py in __init__(self, rows, columns, ordered, **params)
   1313         self.columns = columns
   1314         self.ordered = ordered
-> 1315         super(DataFrame,self).__init__(pdDFrame, allow_None=True, **params)
   1316         self._check_value(self.default)
   1317 

~/ers/gitHub/param/master/param/param/__init__.py in __init__(self, class_, default, instantiate, is_instance, **params)
   1150         self.is_instance = is_instance
   1151         super(ClassSelector,self).__init__(default=default,instantiate=instantiate,**params)
-> 1152         self._check_value(default)
   1153 
   1154 

~/ers/gitHub/param/master/param/param/__init__.py in _check_value(self, val, obj)
   1348                 raise ValueError(msg.format(found=list(val.columns), expected=sorted(self.columns)))
   1349         else:
-> 1350             self._length_bounds_check(self.columns, len(val.columns), 'Column')
   1351 
   1352         if self.ordered:

AttributeError: 'NoneType' object has no attribute 'columns'
```
  
And  
  
`foo = param.DataFrame(rows=3)`  
  
results in:  
  
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-eff4207c7467> in <module>
      5 # class test(param.Parameterized):
      6 #     df = param.DataFrame(rows=3, columns=4)
----> 7 foo = param.DataFrame(rows=3)

~/ers/gitHub/param/master/param/param/__init__.py in __init__(self, rows, columns, ordered, **params)
   1313         self.columns = columns
   1314         self.ordered = ordered
-> 1315         super(DataFrame,self).__init__(pdDFrame, allow_None=True, **params)
   1316         self._check_value(self.default)
   1317 

~/ers/gitHub/param/master/param/param/__init__.py in __init__(self, class_, default, instantiate, is_instance, **params)
   1150         self.is_instance = is_instance
   1151         super(ClassSelector,self).__init__(default=default,instantiate=instantiate,**params)
-> 1152         self._check_value(default)
   1153 
   1154 

~/ers/gitHub/param/master/param/param/__init__.py in _check_value(self, val, obj)
   1356 
   1357         if self.rows is not None:
-> 1358             self._length_bounds_check(self.rows, len(val), 'Row')
   1359 
   1360     def __set__(self,obj,val):

TypeError: object of type 'NoneType' has no len()
```

**Comment by kcpevey** [comment]
Just fyi, I'm working off of the ? magic for docs, https://param.pyviz.org is down

**Comment by jbednar** [comment]
foo=param.DataFrame(default=df, rows=3)

**Comment by jbednar** [comment]
For some reason, a ClassSelector accepts the class as the positional argument, not the default value, a convention which param.DataFrame copies.  We should probably have param.DataFrame pass the first positional argument as the default value, which will make more sense to users. @jlstevens?

**Comment by kcpevey** [comment]
agh. of course! that makes sense given that I know how to other params work.   
  
Given @jbednar 's comment, I'd say that would make more sense to me as a user, but I'd probably also say the same thing about all the param objects.

**Comment by jbednar** [comment]
Changing other parameters' positional arguments would be difficult, but because param.DataFrame has no positional argument, adding one here should be no problem at all and would make it make sense. 

Looking around, most Parameters already do accept the default value as the first argument; the ones that don't are Dict, Array, DataFrame, and Series, ClassSelector, and Composite.  Luckily, most of those don't have any positional arguments, so I think we can and should add positional arguments for Dict, Array, DataFrame, and Series.

**Comment by jbednar** [comment]
Looking at it now, I can see why ClassSelector accepts the class and not the default as the positional argument, which is because for that Parameter the class is always required and the default is not.  Generally positional arguments are done that way, with the most-commonly-required ones before any optional ones, but in this case it's very confusing that the order of arguments differs from the other Parameters.  So if I had to do it over, I'd reorder ClassSelector's signature, but it's been like that for many years, so I'm not going to change that now.

And Composite doesn't take a default value as its positional argument, but that's because it doesn't have any default value per se, and its positional argument is arguably similar to a default value anyway.  So I think Composite is already fine, and after Jean-Luc fixes the other four above then everything will be consistent except ClassSelector.  Should be clearer for everyone except ClassSelector users!



is already correct, I think, as it doesn't actually need a default value, but it does always need a class

**Comment by jlstevens** [comment]
I've assigned myself as I'm happy updating how these parameters process the first (positional) argument.

**Comment by jbednar** [comment]
I might have missed one, but from what I can see, once Jean-Luc is done they should all accept the default value for the first argument except for ClassSelector.

**Comment by kcpevey** [comment]
Sounds good to me!

**Comment by jlstevens** [comment]
@ceball I think `ClassSelector` still requires `default` but the types updated in this PR no longer do. In other words #91 is largely addressed and I think @jbednar has an argument that this is the correct behavior for `ClassSelector`...

**Comment by jbednar** [comment]
I think #91 is largely addressed and can be closed.  From #91, it looks like we also haven't added a positional argument for param.Dynamic, which is not something I would have expected someone to instantiate directly as is done in #91, but maybe for consistency should also have a default positional argument?

To clarify my position on ClassSelector, I think it's *correct* already, but I still think that in retrospect I would have done it differently, as it's not worth the confusion of having this one single Parameter that behaves differently.  We can still consider changing that for param 2.0, but it can't be done before then for backwards compatibility.

**Comment by ceball** [comment]
Closed in favor of #393
